xapian-core/cluster/kmeans.cc

```diff
@@ -24,120 +24,126 @@
+KMeans::initialize_points(const MSet &source) {
+    LOGCALL_VOID(API, "KMeans::initialize_points()", source);
+    TermListGroup tlg(source);
     for (MSetIterator it = source.begin(); it != source.end(); ++it) {
 	Point p;
 	p.initialize(tlg, it.get_document());
 	docs.push_back(p);
     }
 }
 
-bool
-KMeans::converge(vector<Centroid> &previous, vector<Centroid> &current) const {
-    LOGCALL(API, bool, "KMeans::converge()", previous | current);
-    CosineDistance cosine;
-    for (unsigned int i = 0; i < k; ++i) {
-	Centroid &prev = previous[i];
-	Centroid &curr = current[i];
-	double dist = cosine.similarity(prev, curr);
-	if (dist > EPSILON)
-	    return false;
-    }
-    return true;
-}
-
 ClusterSet
 KMeans::cluster(MSet &mset) {
     LOGCALL(API, ClusterSet, "KMeans::cluster()", mset);
     unsigned int size = mset.size();
+    if (k >= size)
+	k = size;
     mset.fetch();
-    if (size <= 0)
-	throw AssertionError("Size of MSet should be greater than zero");
-    if (k <= 0)
-	throw AssertionError("Number of clusters should be greater than zero");
-    if (k > size)
-	throw AssertionError("The number of clusters cannot be greater than number of documents in MSet");
+    initialize_points(mset);
     ClusterSet cset;
-    TermListGroup tlg;
-    tlg.add_documents(mset);
-    CosineDistance cosine;
-    initialize_points(mset, tlg);
-    initialize_centroids(cset);
-    double min = INT_MAX;
-    double min_cluster = 0;
-
-    vector<Centroid> prev;
-    for (int i = 0; i < max_iters; ++i) {
+    initialize_clusters(cset);
+    for (unsigned int i = 0; i < max_iters; ++i) {
+	/** Assign each point to ihe cluster corresponding to its
+	 *  closest cluster centroid
+	 */
+	cset.clear_clusters();
+	CosineDistance distance;
 	for (unsigned int j = 0; j < size; ++j) {
-	    Point &x = docs[j];
-	    min_cluster = 0;
-	    min = INT_MAX;
+	    double closest_cluster_distance = INT_MAX;
+	    double closest_cluster = 0;
 	    for (unsigned int c = 0; c < k; ++c) {
-		Centroid &centroid = centroids[c];
-		double dist = cosine.similarity(x, centroid);
-		if (dist < min) {
-		    min = dist;
-		    min_cluster = c;
+		double dist = distance.similarity(docs[j], cset[c].get_centroid());
+		if (closest_cluster_distance > dist) {
+		    closest_cluster_distance = dist;
+		    closest_cluster = c;
 		}
 	    }
-	    cset.add_to_cluster(x, min_cluster);
+	    cset.add_to_cluster(docs[j], closest_cluster);
 	}
-	prev.clear();
-	for (vector<Centroid>::iterator cit = centroids.begin(); cit != centroids.end(); ++cit)
-	    prev.push_back(*cit);
 
-	cset.recalculate_centroids();
-	centroids.clear();
-	for (unsigned int m = 0; m < k; ++m)
-	    centroids.push_back(cset[m].get_centroid());
+	/** Remember the previous centroids
+	 */
+	previous_centroids.clear();
+	for (unsigned int j = 0; j < k; ++j)
+	    previous_centroids.push_back(cset[j].get_centroid());
 
-	if (converge(prev, centroids))
-	   break;
+	/** Recalculate the centroids for current iteration
```

Comment:
I think all these internal comments can just be `/* … */`, or even single-line `//`. They aren't doc comments.